lib/srv/db/sqlserver/kinit/kinit.go

```diff
@@ -0,0 +1,328 @@
+// Copyright 2022 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package kinit provides utilities for interacting with a KDC (Key Distribution Center) for Kerberos5
+package kinit
+
+import (
+	"bytes"
+	"context"
+	"crypto/x509"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"text/template"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jcmturner/gokrb5/v8/credentials"
+	"github.com/sirupsen/logrus"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth/windows"
+)
+
+const (
+	// krb5ConfigEnv sets the location from which kinit will attempt to read a configuration value
+	krb5ConfigEnv = "KRB5_CONFIG"
+	// kinitBinary is the binary Name for the kinit executable
+	kinitBinary = "kinit"
+	// krb5ConfigTemplate is a configuration template suitable for x509 configuration; it is read by the kinit binary
+	krb5ConfigTemplate = `[libdefaults]
+ default_realm = {{ .RealmName }}
+ rdns = false
+
+
+[realms]
+ {{ .RealmName }} = {
+  kdc = {{ .KDCHostName }}
+  admin_server = {{ .AdminServerName }}
+  pkinit_eku_checking = kpServerAuth
+  pkinit_kdc_hostname = {{ .KDCHostName }}
+ }`
+	// certTTL is the certificate time to live; 1 hour
+	certTTL = time.Minute * 60
+)
+
+// Provider is a kinit provider capable of producing a credentials cacheData for kerberos
+type Provider interface {
+	// UseOrCreateCredentials uses or updates an existing cacheData or creates a new one
+	UseOrCreateCredentials(ctx context.Context) (cache *credentials.CCache, err error)
+}
+
+// PKInit is a structure used for initializing a kerberos context
+type PKInit struct {
+	provider Provider
+}
+
+// UseOrCreateCredentialsCache uses or creates a credentials cacheData.
+func (k *PKInit) UseOrCreateCredentialsCache(ctx context.Context) (*credentials.CCache, error) {
+	return k.provider.UseOrCreateCredentials(ctx)
+}
+
+// New returns a new PKInit initializer
+func New(provider Provider) *PKInit {
+	return &PKInit{provider: provider}
+}
+
+// NewCommandLineInitializerWithCommand returns a new command line initializer using a preinstalled `kinit` binary
+func NewCommandLineInitializerWithCommand(authClient windows.AuthInterface, user, realm, kdcHost, adminServer, dataDir string, ldapCA *x509.Certificate, command CommandGenerator, certGetter CertGetter) *CommandLineInitializer {
```

Comment:
nit:
The `NewCommandLineInitializerWithCommand` signature contains many args. Can we move it to a config struct ?

Reply:
Yeah I had a feeling you'd say that. It started a little smaller and grew a bit too much. Will do.